src/main/java/com/google/sps/ImageTextDectector.java

```diff
@@ -16,45 +16,109 @@
     this.textDetectionAPI = textDetectionAPI;
   }
 
-  public String extractShoppingList(byte[] shoppingImageBytes)
+  public List<String> extractShoppingList(byte[] shoppingImageBytes)
       throws IOException, PhotoDetectionException {
     List<ShoppingListTextEntry> shoppingListText = textDetectionAPI.detect(shoppingImageBytes);
 
-    return createShoppingListQuery(shoppingListText);
+    return createShoppingListQueries(shoppingListText);
   }
 
-  /** Creates query from the text detected by cloudVision API. */
-  private String createShoppingListQuery(List<ShoppingListTextEntry> shoppingListText)
+  /** Creates query from the text detected by Cloud Vision API. */
+  private List<String> createShoppingListQueries(List<ShoppingListTextEntry> shoppingListText)
       throws PhotoDetectionException {
-    // ToDo: Make an algorithm to create query sentences by separating out text returned by
-    // cloudVisionAPI; to group shoppping items based on their position (y axis).
     if (shoppingListText.isEmpty()) {
       throw new PhotoDetectionException("Shopping List doesn't contain any text");
     }
 
-    String queryItem = "";
-    for (ShoppingListTextEntry singleWord : shoppingListText) {
-      queryItem += singleWord.getText() + " ";
+    // Remove the first element of this list because cloud vision API returns all the 
+    // text detetcted as the first element of the list followed by list of single 
+    // words and their properties. Examples of text returned from API can be seen in the test file.
+    shoppingListText.remove(0);
+
+    // To group shoppping items based on their position.
+    int yAxisCurrentLower, yAxisCurrentUpper;
+    String sentence = "";
+    List<String> shoppingQueries = new ArrayList<>();
+
+    // Initialize yAxisPrevUpper with the current upper y value.
+    int yAxisPrevUpper = shoppingListText.get(0).getUpperYBoundary();
+
+    for (ShoppingListTextEntry detectedWord : shoppingListText) {
+      yAxisCurrentLower = detectedWord.getLowerYBoundary();
+      yAxisCurrentUpper = detectedWord.getUpperYBoundary();
+
+      if (isInSameLine(yAxisCurrentLower, yAxisPrevUpper)) {
+        sentence += detectedWord.getText() + " ";
+      } else {
+        shoppingQueries = formatAndAddQuery(sentence, shoppingQueries);
+        sentence = detectedWord.getText() + " ";
+      }
+
+    // Assign current word's upper y value to yAxisPrevUpper for comparisions in the next iteration. 
+      yAxisPrevUpper = yAxisCurrentUpper;
+    }
+
+    shoppingQueries = formatAndAddQuery(sentence, shoppingQueries);
+    return shoppingQueries;
+  }
+
+  private boolean isInSameLine(int currentLowerY, int prevUpperY) {
+    // Compare the lower boundary of the current word box to the upper boundary of the previous word box.
+    // The Cloud Vision API returns words from left to right and from top to bottom
+    // but while giving their cordinates it places sentences/ lines from bottom to top.
```

Comment:
I don't understand this part. Where is the origin of the coordinates you get? Is it top-left or bottom-left?

I think you should reconsider which coordinates you call upper and lower because the logic seems to be reversed here.

Reply:
I mentioned its answer in the 4th review comment (from above).
Please let me know if anything.

Reply:
After looking at the example output, I think the origin is in the top-left corner (which is the standard for computer images). The point you call `lowerY` is actually the y-coordinate of the *upper* left corner and the what you call `upperY` is y-coordinate of the *lower* right corner. The API returns words from top-left to bottom-right with correct coordinates, nothing weird is happening here.

So the solution would be to swap `upper` and `lower` in your naming scheme and update comments.

Does this make sense?

Reply:
Thank you for pointing out. I was thinking of coordinates in terms of mathematics. I understood now. I'll change the names for lower and upper coordinate variables and also update the comments.